pkg/batcher/batcher.go

```diff
@@ -0,0 +1,199 @@
+//
+// The returned Result slice is expected to match the len of the input slice and be in the
+// same order, if order matters for the batched API
+type BatchExecutor[T input, U output] func(ctx context.Context, firstCopiedInput T, input []*T) []Result[U]
+
+// RequestHasher is a function that hashes input to bucket inputs into distinct batches
+type RequestHasher[T input] func(ctx context.Context, input *T) uint64
+
+// NewBatcher creates a batcher that can batch a particular input and output type
+func NewBatcher[T input, U output](ctx context.Context, options Options[T, U]) *Batcher[T, U] {
+	b := &Batcher[T, U]{
+		ctx:      ctx,
+		requests: map[uint64][]*request[T, U]{},
+		trigger:  make(chan struct{}),
+		options:  options,
+	}
+	go b.run()
+	return b
+}
+
+// Add will add an input to the batcher using the batcher's hashing function
+func (b *Batcher[T, U]) Add(ctx context.Context, input *T) Result[U] {
+	request := &request[T, U]{
+		ctx:   ctx,
+		hash:  b.options.RequestHasher(ctx, input),
+		input: input,
+		// The requestor channel is buffered to ensure that the exec runner can always write the result out preventing
+		// any single caller from blocking the others. Specifically since we register our request and then trigger, the
+		// request may be processed while the triggering blocks.
+		requestor: make(chan Result[U], 1),
+	}
+	b.mu.Lock()
+	b.requests[request.hash] = append(b.requests[request.hash], request)
+	b.mu.Unlock()
+	b.trigger <- struct{}{}
+	return <-request.requestor
+}
+
+// DefaultHasher will hash the entire input
+func DefaultHasher[T input](ctx context.Context, input *T) uint64 {
+	hash, err := hashstructure.Hash(input, hashstructure.FormatV2, &hashstructure.HashOptions{SlicesAsSets: true})
+	if err != nil {
+		panic("error hashing")
+	}
+	return hash
+}
+
+// OneBucketHasher will return a constant hash and should be used when there is only one type of request
+func OneBucketHasher[T input](ctx context.Context, input *T) uint64 {
+	return 0
+}
+
+func (b *Batcher[T, U]) run() {
+	for {
+		select {
+		// context that we started with has completed so the app is shutting down
+		case <-b.ctx.Done():
+			return
+		case <-b.trigger:
+			// wait to start the batch of create fleet calls
+		}
+		b.waitForIdle()
+		b.runCalls()
+	}
+}
+
+func (b *Batcher[T, U]) waitForIdle() {
+	timeout := time.NewTimer(b.options.MaxTimeout)
+	idle := time.NewTimer(b.options.IdleTimeout)
+	count := 0
+	for {
+		select {
+		case <-b.trigger:
+			count++
+			if b.options.MaxItems != 0 && count >= b.options.MaxItems {
+				return
+			}
+			if !idle.Stop() {
+				<-idle.C
+			}
+			idle.Reset(b.options.IdleTimeout)
+		case <-timeout.C:
+			return
+		case <-idle.C:
+			return
+		}
+	}
+}
+
+func (b *Batcher[T, U]) runCalls() {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	requestMap := b.requests
+	b.requests = map[uint64][]*request[T, U]{}
+
+	for _, requestBatch := range requestMap {
+		firstInput, err := deepCopy(requestBatch[0].input)
```

Comment:
Why do we need to copy here?

Reply:
the input is modified and used as the base in all implemented batchers. This is kind of awkward to pass thru, but it's either this or in every implemented batcher, the first op is to deepCopy the first request in the batch.

Reply:
I'm not grokking this. Slack?

Reply:
discussed offline, will just mutate the request the input since it's not currently relied on anywhere and probably shouldn't be.